hbase-server/src/test/java/org/apache/hadoop/hbase/replication/TestReplicationWithWALExtendedAttributes.java

```diff
@@ -0,0 +1,295 @@
+
+    htable1.batch(Collections.singletonList(put2), new Object[1]);
+
+    assertGetValues(new Get(ROW), ROW);
+    assertGetValues(new Get(ROW2), ROW2);
+  }
+
+  private static void assertGetValues(Get get, byte[] value)
+    throws IOException, InterruptedException {
+    for (int i = 0; i < NB_RETRIES; i++) {
+      if (i == NB_RETRIES - 1) {
+        fail("Waited too much time for put replication");
+      }
+      Result res = htable2.get(get);
+      if (res.isEmpty()) {
+        LOG.info("Row not available");
+        Thread.sleep(SLEEP_TIME);
+      } else {
+        assertArrayEquals(value, res.value());
+        break;
+      }
+    }
+  }
+
+  public static class TestCoprocessorForWALAnnotationAtSource
+    implements RegionCoprocessor, RegionObserver {
+
+    @Override
+    public Optional<RegionObserver> getRegionObserver() {
+      return Optional.of(this);
+    }
+
+    @Override
+    public void preWALAppend(ObserverContext<RegionCoprocessorEnvironment> ctx, WALKey key,
+      WALEdit edit) throws IOException {
+      key.addExtendedAttribute("extendedAttr1", Bytes.toBytes("Value of Extended attribute 01"));
+      key.addExtendedAttribute("extendedAttr2", Bytes.toBytes("Value of Extended attribute 02"));
+    }
+  }
+
+  public static class TestCoprocessorForWALAnnotationAtSink
+    implements RegionCoprocessor, RegionObserver {
+
+    @Override
+    public Optional<RegionObserver> getRegionObserver() {
+      return Optional.of(this);
+    }
+
+    @Override
+    public void prePut(ObserverContext<RegionCoprocessorEnvironment> c, Put put, WALEdit edit)
+      throws IOException {
+      String attrVal1 = Bytes.toString(put.getAttribute("extendedAttr1"));
+      String attrVal2 = Bytes.toString(put.getAttribute("extendedAttr2"));
+      if (attrVal1 == null || attrVal2 == null) {
+        throw new IOException("Failed to retrieve WAL annotations");
+      }
+      if (
+        attrVal1.equals("Value of Extended attribute 01")
+          && attrVal2.equals("Value of Extended attribute 02")
+      ) {
+        return;
+      }
+      throw new IOException("Failed to retrieve WAL annotations..");
+    }
+
+    @Override
+    public void preBatchMutate(ObserverContext<RegionCoprocessorEnvironment> c,
+      MiniBatchOperationInProgress<Mutation> miniBatchOp) throws IOException {
+      String attrVal1 = Bytes.toString(miniBatchOp.getOperation(0).getAttribute("extendedAttr1"));
+      String attrVal2 = Bytes.toString(miniBatchOp.getOperation(0).getAttribute("extendedAttr2"));
+      if (attrVal1 == null || attrVal2 == null) {
+        throw new IOException("Failed to retrieve WAL annotations");
+      }
+      if (
+        attrVal1.equals("Value of Extended attribute 01")
+          && attrVal2.equals("Value of Extended attribute 02")
+      ) {
+        return;
+      }
+      throw new IOException("Failed to retrieve WAL annotations..");
+    }
+  }
+
+  public static final class TestReplicationSinkRegionServerEndpoint
```

Comment:
@Apache9 @apurtell Even with this approach, we at least need config change to allow adding new class value to `hbase.coprocessor.regionserver.classes` config.

Reply:
Sure, we have a new coprocessor, so it must be registered. This is not a new configuration key-value, though. hbase.coprocessor.regionserver.classes already exists.

Reply:
Sounds good, better than a new config key-value pair to toggle the behavior.